File: src/Core/StagingFinancialEvents/Services/Transformers/LotProcessors/DerivativeLotProcessor.py
from typing import Sequence

import arrow as ar

import Core.FinancialEvents.Schemas.CommonFormats as cf
import Core.FinancialEvents.Schemas.FinancialIdentifier as cfi
import Core.FinancialEvents.Schemas.Grouping as pgf
from Core.StagingFinancialEvents.Contracts.LotProcessor import LotProcessor
from Core.StagingFinancialEvents.Schemas.Lots import StagingTaxLot


class DerivativeLotProcessor(
    LotProcessor[
        StagingTaxLot,
        pgf.TaxLotDerivative,
        Sequence[pgf.TradeEventDerivativeAcquired | pgf.TradeEventDerivativeSold],
    ]
):

    def process(
        self,
        input: StagingTaxLot,
        references: Sequence[pgf.TradeEventDerivativeAcquired | pgf.TradeEventDerivativeSold],
    ) -> pgf.TaxLotDerivative:
        # print("Processing stock lot (ID: {})".format(lot.ID))

        allBuys: list[pgf.TradeEventDerivativeAcquired] = list(filter(lambda trade: isinstance(trade, pgf.TradeEventDerivativeAcquired), references))  # type: ignore
        allSells: list[pgf.TradeEventDerivativeSold] = list(filter(lambda trade: isinstance(trade, pgf.TradeEventDerivativeSold), references))  # type: ignore

        # TODO: Validate returns since buys and sells are merged
        # TODO: What to do when no match is found?
        try:
            matchingBuyById = self.utils.findStockEventById(input.Acquired.ID or "", allBuys)
            matchingSoldByDate = self.utils.findStockEventByDate(input.Sold.DateTime or ar.get("1-0-0"), allSells)[0]

            # print("Matched Buy with trade (ID: {}, DateTime: {})".format(matchingBuyById.ID, matchingBuyById.Date))
            # print("Matched Sell with trade (ID: {}, DateTime: {})".format(matchingSoldByDate.ID, matchingSoldByDate.Date))
        except StopIteration:
            print(
                "Failed processing stock lot (ID: {}, FinancialIdentifier: {}), found no match".format(input.ID, input.FinancialIdentifier)
            )
            raise StopIteration

        processed = pgf.TaxLotDerivative(
            ID=input.ID,
            FinancialIdentifier=cfi.FinancialIdentifier.fromStagingIdentifier(input.FinancialIdentifier),
            Quantity=input.Quantity,
            Acquired=matchingBuyById,
            Sold=matchingSoldByDate,
            ShortLongType=cf.GenericShortLong.LONG,
        )

        return processed

File: lib/core/stagingProcessor/transformers/DerivativeLotProcessor.ts
import { DateTime } from "luxon";
import { FinancialIdentifier } from "@brrr/core/schemas/FinancialIdentifier.ts";
import { GenericShortLong } from "@brrr/core/schemas/CommonFormats.ts";
import type { TradeEventDerivativeAcquired, TradeEventDerivativeSold } from "@brrr/core/schemas/Events.ts";
import type { TaxLotDerivative } from "@brrr/core/schemas/Lots.ts";
import type { StagingTaxLot } from "@brrr/core/schemas/staging/Lots.ts";

function findById<T extends { id: string }>(id: string, items: T[]): T {
	const found = items.find((x) => x.id === id);
	if (!found) throw new Error(`No item found with id: ${id}`);
	return found;
}

function findByDate<T extends { date: DateTime }>(date: DateTime, items: T[]): T[] {
	return items.filter((x) => x.date.toMillis() === date.toMillis());
}

export function processDerivativeLot(
	input: StagingTaxLot,
	references: (TradeEventDerivativeAcquired | TradeEventDerivativeSold)[],
): TaxLotDerivative {
	const allBuys = references.filter((t): t is TradeEventDerivativeAcquired => t.kind === "DerivativeAcquired");
	const allSells = references.filter((t): t is TradeEventDerivativeSold => t.kind === "DerivativeSold");

	let matchingBuy: TradeEventDerivativeAcquired;
	let matchingSell: TradeEventDerivativeSold;

	try {
		matchingBuy = findById(input.acquired.id ?? "", allBuys);
		const soldDate = input.sold.dateTime ?? DateTime.fromMillis(0);
		const sellMatches = findByDate(soldDate, allSells);
		if (sellMatches.length === 0) throw new Error("No sell found");
		matchingSell = sellMatches[0];
	} catch (_e) {
		console.error(`Failed processing derivative lot (ID: ${input.id}), found no match`);
		throw new Error(`LookupError: Failed to match derivative lot (ID: ${input.id})`);
	}

	return {
		id: input.id,
		financialIdentifier: FinancialIdentifier.fromStagingIdentifier(input.financialIdentifier),
		quantity: input.quantity,
		acquired: matchingBuy,
		sold: matchingSell,
		shortLongType: GenericShortLong.LONG,
		provenance: [],
	};
}
